pkg/cli/email/validate.go

```diff
@@ -0,0 +1,154 @@
+package email
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"net/smtp"
+	"regexp"
+	"strings"
+	"time"
+)
+
+//go:generate go run ../../../hack/email/generate_disposable_domains.go
+
+type (
+	optConfig struct {
+		withRFC522     bool
+		checkMX        bool
+		checkMXTimeout time.Duration
+	}
+
+	Option func(config *optConfig)
+)
+
+const (
+	PortSMTP           = 25
+	PortSMTPSubmission = 587
+)
+
+var (
+	// rfc5322: https://stackoverflow.com/a/201378/5405453.
+	rfc5322            = "(?i)(?:[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*|\"(?:[\\x01-\\x08\\x0b\\x0c\\x0e-\\x1f\\x21\\x23-\\x5b\\x5d-\\x7f]|\\\\[\\x01-\\x09\\x0b\\x0c\\x0e-\\x7f])*\")@(?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\\[(?:(?:(2(5[0-5]|[0-4][0-9])|1[0-9][0-9]|[1-9]?[0-9]))\\.){3}(?:(2(5[0-5]|[0-4][0-9])|1[0-9][0-9]|[1-9]?[0-9])|[a-z0-9-]*[a-z0-9]:(?:[\\x01-\\x08\\x0b\\x0c\\x0e-\\x1f\\x21-\\x5a\\x53-\\x7f]|\\\\[\\x01-\\x09\\x0b\\x0c\\x0e-\\x7f])+)\\])"
+	validRfc5322Regexp = regexp.MustCompile(fmt.Sprintf("^%s*$", rfc5322))
+	smtpPorts          = []int{PortSMTPSubmission, PortSMTP}
+)
+
+// Validate validates the email address format without using regular expressions.
+// It runs checks in the order of cheapest to most expensive.
+func Validate(emailAddress string, options ...Option) error {
+	if _, ok := disposableDomains[emailAddress]; ok {
+		return fmt.Errorf(`disposable address "%s" not allowed`, emailAddress)
+	}
+
+	cfg := &optConfig{}
+	for _, applyOption := range options {
+		applyOption(cfg)
+	}
+
+	if cfg.withRFC522 {
+		found := validRfc5322Regexp.Find([]byte(emailAddress))
+
+		if string(found) != emailAddress {
+			return errors.New("not RFC522 compliant")
+		}
+	}
+
+	// Split the email into two parts: local and domain, separated by '@'
+	parts := strings.Split(emailAddress, "@")
+	if len(parts) != 2 {
+		return errors.New("missing @")
+	}
+
+	localPart, domainPart := parts[0], parts[1]
+
+	// Check if both the local part and domain part are non-empty
+	if len(localPart) == 0 || len(domainPart) == 0 {
+		return errors.New("missing parts")
+	}
+
+	// Validate the domain part contains a dot and has text on both sides
+	if strings.Count(domainPart, ".") == 0 || strings.HasPrefix(domainPart, ".") || strings.HasSuffix(domainPart, ".") {
+		return errors.New(`missing "."`)
```

Comment:
Not sure if I understood this correctly, but it seems that the comment does not match the implementation.
If it's about the dots should it be something like this:

```suggestion
  if strings.Count(domainPart, ".") == 0 {
      return errors.New(`missing "."`)
  }
  if strings.HasPrefix(domainPart, ".") || strings.HasSuffix(domainPart, ".") {
      return errors.New(`"." cannot be at the start or end of domain`)
  }
```